File: Documentation/4.2/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Querying/HowToWorkWithSuggestions.cs
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Raven.Client.Documents;
using Raven.Client.Documents.Indexes;
using Raven.Client.Documents.Queries.Suggestions;
using Raven.Documentation.Samples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Session.Querying
{
    public class HowToWorkWithSuggestions
    {
        private class Foo
        {
            #region suggest_7
            public int PageSize { get; set; } = 15;

            public StringDistanceTypes? Distance { get; set; } = StringDistanceTypes.Levenshtein;

            public float? Accuracy { get; set; } = 0.5f;

            public SuggestionSortMode SortMode { get; set; } = SuggestionSortMode.Popularity;
            #endregion
        }

        private interface IFoo<T>
        {
            #region suggest_1
            ISuggestionQuery<T> SuggestUsing<T>(SuggestionBase suggestion);

            ISuggestionQuery<T> SuggestUsing<T>(Action<ISuggestionBuilder<T>> builder);
            #endregion

            #region suggest_2
            ISuggestionOperations<T> ByField(string fieldName, string term);

            ISuggestionOperations<T> ByField(string fieldName, string[] terms);

            ISuggestionOperations<T> ByField(Expression<Func<T, object>> path, string term);

            ISuggestionOperations<T> ByField(Expression<Func<T, object>> path, string[] terms);

            ISuggestionOperations<T> WithOptions(SuggestionOptions options);
            #endregion
        }

        private class Employees_ByFullName : AbstractIndexCreationTask<Employee>
        {
        }

        public async Task Sample()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region suggest_5
                    Dictionary<string, SuggestionResult> suggestions = session
                        .Query<Employee, Employees_ByFullName>()
                        .SuggestUsing(builder => builder
                            .ByField("FullName", "johne")
                            .WithOptions(new SuggestionOptions
                            {
                                Accuracy = 0.4f,
                                PageSize = 5,
                                Distance = StringDistanceTypes.JaroWinkler,
                                SortMode = SuggestionSortMode.Popularity
                            }))
                        .Execute();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region suggest_6
                    Dictionary<string, SuggestionResult> suggestions = await asyncSession
                        .Query<Employee, Employees_ByFullName>()
                        .SuggestUsing(builder => builder
                            .ByField("FullName", "johne")
                            .WithOptions(new SuggestionOptions
                            {
                                Accuracy = 0.4f,
                                PageSize = 5,
                                Distance = StringDistanceTypes.JaroWinkler,
                                SortMode = SuggestionSortMode.Popularity
                            }))
                        .ExecuteAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region suggest_8
                    Dictionary<string, SuggestionResult> suggestions = session
                        .Query<Employee, Employees_ByFullName>()
                        .SuggestUsing(new SuggestionWithTerm("FullName") { Term = "johne" })
                        .Execute();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region suggest_9
                    Dictionary<string, SuggestionResult> suggestions = await asyncSession
                        .Query<Employee, Employees_ByFullName>()
                        .SuggestUsing(new SuggestionWithTerm("FullName") { Term = "johne" })
                        .ExecuteAsync();
                    #endregion
                }
            }
        }
    }
}

File: Documentation/5.4/Samples/php/ClientApi/Session/Querying/HowToWorkWithSuggestions.php
<?php

use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Queries\Suggestions\StringDistanceTypes;
use RavenDB\Documents\Queries\Suggestions\SuggestionBase;
use RavenDB\Documents\Queries\Suggestions\SuggestionDocumentQueryInterface;
use RavenDB\Documents\Queries\Suggestions\SuggestionOperationsInterface;
use RavenDB\Documents\Queries\Suggestions\SuggestionOptions;
use RavenDB\Documents\Queries\Suggestions\SuggestionResult;
use RavenDB\Documents\Queries\Suggestions\SuggestionSortMode;
use RavenDB\Documents\Queries\Suggestions\SuggestionWithTerm;
use RavenDB\Samples\Infrastructure\DocumentStoreHolder;
use RavenDB\Type\StringArray;

class Foo {
    # region suggest_7
    private int $pageSize = 15;
    private ?StringDistanceTypes $distance = null;
    private float $accuracy = 0.5;
    private ?SuggestionSortMode $sortMode = null;

    public function __construct()
    {
        $distance = StringDistanceTypes::levenshtein();
        $sortMode = SuggestionSortMode::popularity();
        ...
    }

    // getters and setters for fields listed above
    # endregion
}

interface FooInterface {
    # region suggest_1
    /**
     * Usage:
     *   - suggestUsing(SuggestionBase $suggestion);
     *   - suggestUsing(Closure $suggestionBuilder);
     *
     * @param SuggestionBase|Closure|null $suggestionOrBuilder
     * @return SuggestionDocumentQueryInterface
     */
    public function suggestUsing(null|SuggestionBase|Closure $suggestionOrBuilder): SuggestionDocumentQueryInterface;

    # endregion

    # region suggest_2
    /**
     * Usage:
     *   - byField("fieldName", "term");
     *   - byField("fieldName", ["term1", "term2"]);
     */
    function byField(?string $fieldName, null|string|StringArray|array $terms): SuggestionOperationsInterface;

    function withOptions(?SuggestionOptions $options): SuggestionOperationsInterface;
    # endregion
}

class Employees_ByFullName extends AbstractIndexCreationTask
{

}

class Employee {

}

class HowToWorkWithSuggestions
{
    public function examples(): void
    {
        $store = DocumentStoreHolder::getStore();
        try {
            $session = $store->openSession();
            try {
                # region suggest_5
                $options = new SuggestionOptions();
                $options->setAccuracy(0.4);
                $options->setPageSize(5);
                $options->setDistance(StringDistanceTypes::jaroWinkler());
                $options->setSortMode(SuggestionSortMode::popularity());

                /** @var array<SuggestionResult> $suggestions */
                $suggestions = $session
                    ->query(Employee::class, Employees_ByFullName::class)
                    ->suggestUsing(function($builder) use ($options) {
                        $builder->byField("FullName", "johne")
                            ->withOptions($options);
                    })
                    ->execute();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region suggest_8
                $suggestionWithTerm = new SuggestionWithTerm("FullName");
                $suggestionWithTerm->setTerm("johne");

                /** @var array<SuggestionResult> $suggestions */
                $suggestions = $session
                    ->query(Employee::class, Employees_ByFullName::class)
                    ->suggestUsing($suggestionWithTerm)
                    ->execute();
                # endregion
            } finally {
               $session->close();
            }
        } finally {
            $store->close();
        }
    }
}
